bundles/test/org.eclipse.smarthome.magic/src/main/java/org/eclipse/smarthome/magic/binding/internal/firmware/MagicFirmwareProvider.java

```diff
@@ -0,0 +1,102 @@
+/**
+ * Copyright (c) 2014,2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.magic.binding.internal.firmware;
+
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.binding.firmware.Firmware;
+import org.eclipse.smarthome.core.thing.binding.firmware.FirmwareBuilder;
+import org.eclipse.smarthome.core.thing.firmware.FirmwareProvider;
+import org.eclipse.smarthome.magic.binding.MagicBindingConstants;
+import org.osgi.service.component.ComponentContext;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+
+/**
+ * Provides firmware for the magic thing type for firmware update.
+ *
+ * @author Dimitar Ivanov - Initial contribution
+ */
+@Component(service = FirmwareProvider.class)
+public class MagicFirmwareProvider implements FirmwareProvider {
+
+    private final Map<ThingTypeUID, Set<Firmware>> magicFirmwares = new HashMap<ThingTypeUID, Set<Firmware>>();
+
+    /**
+     * So far only this thing type is used for the firmware update feature of the magic binding. If this changes, we can
+     * add a parameter to the {@link #createFirmware(String, String, boolean)} and
+     * {@link #addFirmware(Firmware)} methods.
+     */
+    private final ThingTypeUID thingTypeUID = MagicBindingConstants.THING_TYPE_FIRMWARE_UPDATE;
```

Comment:
This should be a constant

Reply:
Yes, you are right. I would even prefer to use the `MagicBindingConstants.THING_TYPE_FIRMWARE_UPDATE` everywhere, because it does make sense to define extra constant here.